rust/lance-encoding/src/encodings/physical/dictionary.rs

```diff
@@ -154,6 +199,81 @@ impl PrimitivePageDecoder for DictionaryPageDecoder {
     }
 }
 
+/// An encoder for data that is already dictionary encoded.  Stores the
+/// data as a dictionary encoding.
+#[derive(Debug)]
+pub struct AlreadyDictionaryEncoder {
+    indices_encoder: Box<dyn ArrayEncoder>,
+    items_encoder: Box<dyn ArrayEncoder>,
+}
+
+impl AlreadyDictionaryEncoder {
+    pub fn new(
+        indices_encoder: Box<dyn ArrayEncoder>,
+        items_encoder: Box<dyn ArrayEncoder>,
+    ) -> Self {
+        Self {
+            indices_encoder,
+            items_encoder,
+        }
+    }
+}
+
+impl ArrayEncoder for AlreadyDictionaryEncoder {
+    fn encode(&self, arrays: &[ArrayRef], buffer_index: &mut u32) -> Result<EncodedArray> {
+        // If we get multiple dictionary arrays and they don't all have the same dictionary
+        // then we need to normalize the indices.  Otherwise we might have something like:
+        //
+        // First chunk ["hello", "foo"], [0, 0, 1, 1, 1]
+        // Second chunk ["bar", "world"], [0, 1, 0, 1, 1]
+        //
+        // If we simply encode as ["hello", "foo", "bar", "world"], [0, 0, 1, 1, 1, 0, 1, 0, 1, 1]
+        // then we will get the wrong answer.
+        //
+        // A simple way to do this today is to just concatenate all the arrays.
+        //
+        // TODO: We could be more efficient here by checking if the dictionaries are the same
+        //       Also, if they aren't, we can possibly do something cheaper than concatenating
+        let array_refs = arrays.iter().map(|arr| arr.as_ref()).collect::<Vec<_>>();
+        let array = arrow_select::concat::concat(&array_refs)?;
```

Comment:
Nice that there's this util for normalizing the indices (It looks like that is what this is doing?)